### ISSUE TITLE
evaluation for expressions doesn't seem to work when setting a variable version

### ISSUE THREAD

**Issue by u-de-o** [iabudiab/dependency-track-maven-plugin#20]
Hi,

we are having the following setup in our pom.xml:

### Setup

```
<plugin>
    <groupId>org.codehaus.mojo</groupId>
    <artifactId>build-helper-maven-plugin</artifactId>
    <executions>
        <execution>
            <id>parse-version</id>
            <goals>
                <goal>parse-version</goal>
            </goals>
            <configuration>
                <propertyPrefix>app.version</propertyPrefix>
            </configuration>
        </execution>
    </executions>
</plugin>
```

This plugin is being executed in the phase **validate** of the maven build.
It generates among others a variable called `parsedVersion.majorVersion`, which expresses the major version of a project's version:

```
[DEBUG] Loading mojo org.codehaus.mojo:build-helper-maven-plugin:3.5.0:parse-version from plugin realm ClassRealm[plugin>org.codehaus.mojo:build-helper-maven-plugin:3.5.0, parent: jdk.internal.loader.ClassLoaders$AppClassLoader@277050dc]
[DEBUG] Configuring mojo execution 'org.codehaus.mojo:build-helper-maven-plugin:3.5.0:parse-version:parse-version' with basic configurator -->
[...]
[DEBUG]   (s) propertyPrefix = parsedVersion
[DEBUG]   (f) qualifierPrefix = -
[DEBUG]   (f) versionString = 1.29-SNAPSHOT
[DEBUG] -- end configuration --
[DEBUG] Parsed Version
[DEBUG]          major: 1
[DEBUG]          minor: 29
[DEBUG]    incremental: 0
[DEBUG]    buildnumber: 0
[DEBUG]      qualifier: SNAPSHOT
[DEBUG] define property parsedVersion.majorVersion = "1"
[DEBUG] define property parsedVersion.minorVersion = "29"
[DEBUG] define property parsedVersion.incrementalVersion = "0"
[DEBUG] define property parsedVersion.buildNumber = "0"
[...]
```

We are also using the following:

```
<plugin>
    <groupId>dev.iabudiab</groupId>
    <artifactId>dependency-track-maven-plugin</artifactId>
    <configuration>
        <dependencyTrackUrl>SOME_URL</dependencyTrackUrl>
        <projectName>SOME_NAME</projectName>
        <projectVersion>${parsedVersion.majorVersion}</projectVersion>
        // security gates
    </configuration>
</plugin>
```

We are later then invoking everything by `mvn org.cyclonedx:cyclonedx-maven-plugin:makeAggregateBom` and `mvn --projects . dev.iabudiab:dependency-track-maven-plugin:upload-bom -DskipTests -DdependencyTrackApiKey=SOME_API_KEY`.

### Problem

No matter what we try, the generated version from the maven build-helper is not being resolved to the version to be used when uploading the bom.
It always falls back to the default (${project.version}).

I checked some Mojos of your project and it seems the default phase is verify.
So, if you execute `mvn verify help:evaluate -Dexpression=parsedVersion.majorVersion -DskipTests` you can see the output is `1` as it should be for version `1.29-SNAPSHOT`, so the result of the variable is there.

Is this a bug or the wanted behaviour?

Thank you in advance!

**Comment by u-de-o** [comment]
I found the real issue by some code inspection.

The value from `projectVersion` is just a `@Parameter(property = "projectVersion", defaultValue = "${project.version}", required = true)`, that is being forwarded to the dependency-track-API without further validation or evaluation.

The documentation of `org.apache.maven.plugins.annotations.Parameter`states, that the `defaultValue` can be an expression that is being interpreted at inject time, as also used in your code.
In the description of your pom.xml's parameter `projectVersion` it refers to that default value `${project.version}` and for me that leads to the conclusion, that it's also possible to use an expression in the `projectVersion` itself.

But unfortunately it is not.

It would be very useful to be able to use dynamic versions here. Maybe the PluginParameterExpressionEvaluator mentioned in the documentation could help somehow...

**Comment by iabudiab** [comment]
@Esemudeo Hey there 👋 Thanks for the detailed report and description. Very much appreciated 👍 

If I am not mistaken, then based on your description everything is working as it should.

You are calling the `dependency-track-maven-plugin` separately via `mvn dev.iabudiab:dependency-track-maven-plugin:upload-bom -DskipTests -DdependencyTrackApiKey=SOME_API_KEY`

In this case the plugin is targeted and executed separately and it is not participating in any phase, thus, the results of other plugin executions from the previous phases are not visible here.

To verify this, define an `execution` in the plugin config in the `pom.xml`

```xml
<plugin>
    <groupId>dev.iabudiab</groupId>
    <artifactId>dependency-track-maven-plugin</artifactId>
    <configuration>
        <dependencyTrackUrl>SOME_URL</dependencyTrackUrl>
        <projectName>SOME_NAME</projectName>
        <projectVersion>${parsedVersion.majorVersion}</projectVersion>
        <executions>
              <execution>
                     <phase>verify</phase>
                     <goals>
                            <goal>upload-bom</goal>
                     </goals>
              </execution>
        </executions>
    </configuration>
</plugin>
```

then try to run the `verify` phase, i.e. `mvn verify`

This way, all the plugins up-to the `verify` phase are executed and their results are visible downstream.

I've reproduced your setup locally and with the following setup everything works as you would like:

```xml
<plugin>
    <groupId>org.codehaus.mojo</groupId>
    <artifactId>build-helper-maven-plugin</artifactId>
    <executions>
        <execution>
            <id>parse-version</id>
            <goals>
                <goal>parse-version</goal>
            </goals>
            <configuration>
                <propertyPrefix>app.version</propertyPrefix>
            </configuration>
        </execution>
    </executions>
</plugin>

<plugin>
    <groupId>dev.iabudiab</groupId>
    <artifactId>dependency-track-maven-plugin</artifactId>
    <version>2.4.1</version>
    <executions>
        <execution>
            <phase>verify</phase>
            <goals>
                <goal>upload-bom</goal>
            </goals>
        </execution>
    </executions>
    <configuration>
        <dependencyTrackUrl>API_URL</dependencyTrackUrl>
        <dependencyTrackApiKey>SOME_API_KEY</dependencyTrackApiKey>
        <projectName>SOME_NAME</projectName>
        <projectVersion>${app.version.majorVersion}</projectVersion>
    </configuration>
</plugin>
```

```shell
$ mvn verify
....
[INFO] DependencyTrack Maven Plugin
[INFO] DependencyTrack URL             : API_URL
[INFO] Project name                    : SOME_NAME
[INFO] Project version                 : 2 <=== this is the major version in my case
[INFO] Skip                            : false
[INFO] Using artifact directory        : /<path>/<to>/<artifact>
[INFO] Using artifact                  : bom.xml
[INFO] Upload matching suppressions    : false
[INFO] Reset expired suppressions      : true
[INFO] ProjectMetrics retry delay      : 5
[INFO] ProjectMetrics retry limit      : 3
.....
```

Let me know if it helps or if I misunderstood the problem.

Cheers

**Comment by u-de-o** [comment]
Hi @iabudiab , thank you for your fast reply and detailed explanation.
So it seems the projectVersion in your plugin actually **can** work with expressions, I just need to explicitly run a specific phase in the command.

I will try that tomorrow morning as soon I get back to work.

Thank you very much in advance, I will report if this works for us!

**Comment by iabudiab** [comment]
Hey @Esemudeo, you're welcome!

Yes, resolving expressions is not something specific to this plugin, but rather how all maven plugins work in general. Plugins' results are visible to all participating downstream plugins up-to the given phase. Plugins executed separately, e.g. `mvn org.cyclonedx:cyclonedx-maven-plugin:makeAggregateBom` are just that: separate/standalone/isolated. 

Let me know if you still have a problem or other questions 😉


Update:
To expand upon the idea above and to be more precise: `Plugins' results are visible to all participating downstream plugins up-to the given phase`:
Maven itself doesn't provide any special mechanisms to allow for data flows between plugins. The usual way to achieve this is via files. However, in this case of `build-helper-maven-plugin` it works because the internal `project` model is "shared" during the execution of the plugins.

There are other ways to achieve your goal, if you don't want/can't run the whole `verify` phase at once. For example you could run the following during CI:

```
mvn \
  build-helper:parse-version \
  versions:set -DnewVersion=\${parsedVersion.majorVersion}.\${parsedVersion.nextMinorVersion}

mvn dev.iabudiab:dependency-track-maven-plugin:upload-bom -DskipTests -DdependencyTrackApiKey=SOME_API_KEY
```

PS: I've typed this directly here w/o verifying or running this locally, so it could contain a typo/error or two  ;)

**Comment by iabudiab** [comment]
@Esemudeo  Hey there 👋

Did it work? Can we close this issue or do you still have problems / questions?

**Comment by u-de-o** [comment]
Hi @iabudiab ,

Sorry got hung up in coding again...

Yes, thank you.
With your help I understood how I needed to use certain phases and/or goals to achieve that maven provides me the variables I need.

I just started my command with ``` mvn validate dev.iabudiab ...``` and then had everything I needed.

For others and future problems:

```mvn build-helper-maven-plugin:parse-version dev.iabudiab ...``` also would have helped.
In this way only the ```parse-version``` goal would have been executed instead of the whole maven ```validate``` phase.
The problem with running the whole phase can be ```executions``` defined in related ```pom.xml``` files which trigger automatically and do much more than you'd actually might need.
